x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -72,50 +82,140 @@ public void performAction(IndexMetadata indexMetadata, ClusterState clusterState
                         .masterNodeTimeout(TimeValue.MAX_VALUE)
                         .settings(settings);
-                getClient().admin().indices().updateSettings(updateSettingsRequest,
+                    getClient().admin().indices().updateSettings(updateSettingsRequest,
                         ActionListener.wrap(response -> listener.onResponse(null), listener::onFailure));
-            } else {
-                // No nodes currently match the allocation rules, so report this as an error and we'll retry
-                logger.debug("could not find any nodes to allocate index [{}] onto prior to shrink", indexName);
-                listener.onFailure(new NoNodeAvailableException("could not find any nodes to allocate index [" + indexName + "] onto" +
-                    " prior to shrink"));
-            }
+                } else {
+                    // No nodes currently match the allocation rules or have no enough disk bytes,
+                    // so report this as an error and we'll retry
+                    logger.debug("could not find any nodes to allocate index [{}] onto prior to shrink", indexName);
+                    listener.onFailure(new NoNodeAvailableException("could not find any nodes to allocate index [" + indexName +
+                        "] onto prior to shrink"));
+                }
+            }, (Exception e) -> listener.onFailure(new NoNodeAvailableException("failed to retrieve disk information" +
+                " to select a single node for shard copy allocation"))));
+
         } else {
             // There are no shards for the index, the index might be gone. Even though this is a retryable step ILM will not retry in
             // this case as we're using the periodic loop to trigger the retries and that is run over *existing* indices.
             listener.onFailure(new IndexNotFoundException(indexMetadata.getIndex()));
         }
     }
 
+    Optional<String> selectSingleNode(Set<String> validNodeIds, Map<String, Long> nodeShardsStorageBytes) {
+        List<Map.Entry<String, Long>> nodeShardsStorageList = new ArrayList<>(nodeShardsStorageBytes.entrySet());
+        nodeShardsStorageList.sort((o1, o2) -> o2.getValue().compareTo(o1.getValue()));
+        Optional<String> nodeId = Optional.empty();
+        for (Map.Entry<String, Long> entry : nodeShardsStorageList) {
+            // we prefer to select the node which contains the maximum shards storage bytes of the index from the valid node list
+            if (validNodeIds.contains(entry.getKey())) {
+                nodeId = Optional.of(entry.getKey());
+                break;
+            }
+        }
+
+        // if we cannot find a node which contains any shard of the index,
+        // shuffle the valid node list and select randomly
+        if (nodeId.isEmpty()) {
+            List<String> list = new ArrayList<>(validNodeIds);
+            Randomness.shuffle(list);
+            nodeId = list.stream().findAny();
```

Comment:
since we shuffled above, shall we just get the first one? (to avoid extra allocations by streaming the list)

Reply:
Yeah, I think we should do that, I've changed the code yet.